pkg/agent/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,411 @@
+	supportBundleNodeType controlplane.SupportBundleCollectionNodeType,
+	namespace string,
+	antreaClientGetter agent.AntreaClientProvider,
+	ovsCtlClient ovsctl.OVSCtlClient,
+	aq agentquerier.AgentQuerier,
+	npq querier.AgentNetworkPolicyInfoQuerier,
+	v4Enabled,
+	v6Enabled bool) *SupportBundleController {
+	c := &SupportBundleController{
+		nodeName:              nodeName,
+		supportBundleNodeType: supportBundleNodeType,
+		namespace:             namespace,
+		antreaClientGetter:    antreaClientGetter,
+		queue:                 workqueue.NewNamed("supportbundle"),
+		ovsCtlClient:          ovsCtlClient,
+		aq:                    aq,
+		npq:                   npq,
+		v4Enabled:             v4Enabled,
+		v6Enabled:             v6Enabled,
+		sftpUploader:          &sftpUploader{},
+	}
+	return c
+}
+func (c *SupportBundleController) watchSupportBundleCollections() {
+	klog.Info("Starting watch for SupportBundleCollections")
+	antreaClient, err := c.antreaClientGetter.GetAntreaClient()
+	if err != nil {
+		klog.ErrorS(err, "Failed to get antrea client")
+		return
+	}
+	options := metav1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector("nodeName", c.nodeName).String(),
+	}
+	watcher, err := antreaClient.ControlplaneV1beta2().SupportBundleCollections().Watch(context.TODO(), options)
+	if err != nil {
+		klog.ErrorS(err, "Failed to start watch for SupportBundleCollections")
+		return
+	}
+	// Watch method doesn't return error but "emptyWatch" in case of some partial data errors,
+	// e.g. timeout error. Make sure that watcher is not empty and log warning otherwise.
+	if reflect.TypeOf(watcher) == reflect.TypeOf(emptyWatch) {
+		klog.ErrorS(nil, "Failed to start watch for SupportBundleCollections, please ensure antrea service is reachable for the agent")
+		return
+	}
+
+	klog.Info("Started watch for SupportBundleCollections")
+	eventCount := 0
+	defer func() {
+		klog.InfoS("Stopped watch for SupportBundleCollections", "total items received", eventCount)
+		watcher.Stop()
+	}()
+
+	for {
+		select {
+		case event, ok := <-watcher.ResultChan():
+			if !ok {
+				return
+			}
+			switch event.Type {
+			case watch.Bookmark:
+				klog.V(2).Info("Received Bookmark event")
+			case watch.Added:
+				c.addSupportBundleCollection(event.Object.(*cpv1b2.SupportBundleCollection))
+				klog.InfoS("Added SupportBundleCollection", "name", event.Object.(*cpv1b2.SupportBundleCollection).Name)
+			case watch.Deleted:
+				c.deleteSupportBundleCollection(event.Object.(*cpv1b2.SupportBundleCollection))
+				klog.InfoS("Deleted SupportBundleCollection", "name", event.Object.(*cpv1b2.SupportBundleCollection).Name)
+			default:
+				klog.ErrorS(nil, "Received unknown event", "event", event.Type)
+				return
+			}
+			eventCount++
+		}
+
+	}
+}
+
+func (c *SupportBundleController) addSupportBundleCollection(supportBundle *cpv1b2.SupportBundleCollection) {
+	c.supportBundleCollectionsMutex.Lock()
+	defer c.supportBundleCollectionsMutex.Unlock()
```

Comment:
Unlock the mutex after cosuming resource, no need to wait until it is added into queue? 

```
c.supportBundleCollectionsMutex.Lock()
c.supportBundleCollection = supportBundle
c.supportBundleCollectionsMutex.Unlock()
c.queue.Add(supportBundle.Name)
```

Reply:
updated